### DISCUSSION TITLE
Discretize System Becomes Infeasible

### DISCUSSION THREAD

**Discussion by Vfriday** [yalmip/YALMIP#1134] [Problems using YALMIP] [answered]
I need assistance to get my system working.  I used yalmip to simulate my system in continuous-timeand it worked well but on implementing it in discrete time i began to have infeasible message, this is particularlt=y due to constraint c5 of my codew which is necessary to placing the pole and synthesizing LMI gain.
The code is pasted below:

```
%The Discrete-time model
%state space Equations
A6= [ 0.8852         0   -0.4844         0    0.0543         0         0         0
         0    0.8852         0   -0.4844         0    0.0543         0         0
    0.2102         0    0.9351         0   -0.2112         0   -0.0003   -0.0125
         0    0.2102         0    0.9351         0   -0.2112    0.0125   -0.0003
    0.0106         0    0.0953         0    0.9372    0.0011    0.0081    0.1122
         0    0.0106         0    0.0953   -0.0011    0.9372   -0.1122    0.0081
         0         0    0.0009         0    0.0177   -0.0009    0.9931   -0.0969
         0         0         0    0.0009    0.0009    0.0177    0.0969    0.9931];
B6= [0.0038         0
         0    0.0038
    0.0004         0
         0    0.0004
         0         0
         0         0
         0         0
         0         0];
C6= [16     0     0     0     0     0     0     0
     0    16     0     0     0     0     0     0];
D6 = [ 0     0
     0     0];
E6 = B6;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% LMIs H2-Observer Design
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%Discretize Linear Submodel State-Space Matrices %%%%%%%
yalmip('clear')

%%%%%%%  Strip Stability Domain Parameters %%%%%%%

Alpha = 500;               % minimum real part of eigenvalues
Beta = 600 ;                  % maximum real part of eigenvalues 

% %%%%%%% LMI Design Using YALMIP %%%%%%%%
% % NB : YALMIP and the solver folders must be added to the path
% % https://yalmip.github.io/ to get more infos
% 
ops = sdpsettings('solver','sedumi');   % Choice of solver 
% 
% %%% LMI Variables 
n = size(A6,1); 
m = size(B6,2);
p = size(C6,1);
X = sdpvar(n,n);                % 8x8 symmetrical matrix 
Y = sdpvar(m,n);                       % 2x8 matrix 
Z = sdpvar(p,p);                       % 2x2 symmetrical matrix 

gamma = sdpvar(1,1);                   % Upper bound of H2 norm to minimize
% 
% %%% LMI Constraints
  


c1 = [[-X,X*A6+Y'*C6;A6'*X+C6'*Y,-X]<=0];   %From proposition 4.2 on pg97,remark 4.1
c2 = [[-X,A6'*X+C6'*Y;X*A6+Y'*C6,-X]<=0];   %"""""""""""""""""""""Schur Stabilizability

c3 = [[-X,X*A6;A6'*X,-X-gamma*(C6')*C6]<=0];        %""""""""""""""""" Detectability
c4 = [[-Z,E6'*X;X*E6,-X]<=0];

c5 = [[A6'*X+A6*X+C6'*Y+Y'*C6+2*Alpha*X]<=0]; %Proposition 4.3
c6 = [[A6'*X+A6*X+C6'*Y+Y'*C6+2*Beta*X]>=0];  %""""""""""""""" Vertical Strip Region
c7 = [trace(Z)<=gamma];



st_lmi = c2+c3+c4+c5+c6+c7;         % Constraint input of optimize()

sol=optimize(st_lmi,gamma);    

% % % Analyze error flags
if sol.problem == 0
 % Extract and display value
 solution = value(X)
else
 display('Hmm, something went wrong!');
 sol.info
 yalmiperror(sol.problem)
end
```

**Comment by johanlofberg** [comment, marked as answer]
You have a typo in c5 causing it to lack symmetry

```
>> c5
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
|   ID|                     Constraint|     Coefficient range|
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
|   #1|   Element-wise inequality 64x1|   0.0006 to 1001.9862|
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

```

Same thing in c6

**Reply by Vfriday** [reply]
Thank you Dr. Lofberg for all you are doing for this community and for your prompt response, your legacy will be sustained. I am actually trying to code proposition 4.3  or equation 4.9 of page 99 of Yuan and Dul book: 
P > 0
A'P + PA + 2αP < 0
A'P + PA + 2βP > 0
Please can you help me out with the right coding.

**Reply by johanlofberg** [reply]
Where are you stuck? You just add a multiplication sign and you are done (and dehomogenize with P>=I as P=0 is a trivial solution) https://yalmip.github.io/tutorial/semidefiniteprogramming/

**Reply by Vfriday** [reply]
I coded the proposition that way because I am trying to synthesize a discrete H2 observer, I have been on this for over a month

**Reply by johanlofberg** [reply]
The code you posted, and the LMIs you write later are not the same, so not sure what you are talking about. The bugs in c5 and c6 are obvious, you're not typing in the equations you most likely have on paper as you have weird order on A6 and X.

**Reply by Vfriday** [reply]
I Implemented your advice sir and I am getting  infeasible solution. The implementation is below:
`%The Discrete-time model
%state space Equations
A6= [ 0.8852         0   -0.4844         0    0.0543         0         0         0
         0    0.8852         0   -0.4844         0    0.0543         0         0
    0.2102         0    0.9351         0   -0.2112         0   -0.0003   -0.0125
         0    0.2102         0    0.9351         0   -0.2112    0.0125   -0.0003
    0.0106         0    0.0953         0    0.9372    0.0011    0.0081    0.1122
         0    0.0106         0    0.0953   -0.0011    0.9372   -0.1122    0.0081
         0         0    0.0009         0    0.0177   -0.0009    0.9931   -0.0969
         0         0         0    0.0009    0.0009    0.0177    0.0969    0.9931];
B6= [0.0038         0
         0    0.0038
    0.0004         0
         0    0.0004
         0         0
         0         0
         0         0
         0         0];
C6= [16     0     0     0     0     0     0     0
     0    16     0     0     0     0     0     0];
D6 = [ 0     0
     0     0];
E6 = B6;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% LMIs H2-Observer Design
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%Discretize Linear Submodel State-Space Matrices %%%%%%%
yalmip('clear')

%%%%%%%  Strip Stability Domain Parameters %%%%%%%

Alpha = 500;               % minimum real part of eigenvalues
Beta = 600 ;                  % maximum real part of eigenvalues 

% %%%%%%% LMI Design Using YALMIP %%%%%%%%
% % NB : YALMIP and the solver folders must be added to the path
% % https://yalmip.github.io/ to get more infos
% 
ops = sdpsettings('solver','sedumi');   % Choice of solver 
% 
% %%% LMI Variables 
n = size(A6,1); 
m = size(B6,2);
p = size(C6,1);
X = sdpvar(n,n,'full');                % 8x8 fully parameterized square matrix 
Y = sdpvar(m,n);                       % 2x8 matrix 
Z = sdpvar(p,p);                       % 2x2 symmetrical matrix 
W=sdpvar(size(B6,2) ,size(B6,1),'full');
gamma = sdpvar(1,1);                   % Upper bound of H2 norm to minimize
% 
% %%% LMI Constraints
%   smallnum = 0.00001;
c0 = [X>=1e-5*eye(size(X))];

c1 = [[-X,X*A6+Y'*C6;A6'*X+C6'*Y,-X]<=0];   %From proposition 4.2 on pg97,remark 4.1
c2 = [[-X,A6'*X+C6'*Y;X*A6+Y'*C6,-X]<=0];   %"""""""""""""""""""""Schur Stabilizability

c3 = [[-X,X*A6;A6'*X,-X-gamma*(C6')*C6]<=0];        %""""""""""""""""" Detectability
c4 = [[-Z,E6'*X;X*E6,-X]<=0];

c5 = [[A6'*X+A6*X+2*Alpha*X]<=0];
c6 = [[A6'*X+A6*X+2*Beta*X]>=0];
% c5 = [[A6'*X+X*A6+C6'*Y+Y'*C6+2*Alpha*X]<=0]; %Proposition 4.3
% c6 = [[A6'*X+A6*X+C6'*Y+Y'*C6+2*Beta*X]>=0];  %""""""""""""""" Vertical Strip Region
c7 = [trace(Z)<=gamma];



st_lmi = c0+c1+c2+c3+c4+c5+c6+c7;         % Constraint input of optimize()

sol=optimize(st_lmi,gamma);    

% % % Analyze error flags
if sol.problem == 0
 % Extract and display value
 solution = value(X)
else
 display('Hmm, something went wrong!');
 sol.info
 yalmiperror(sol.problem)
end`

**Reply by johanlofberg** [reply]
Please put code in code blocks (< > icon, as I've done for your first post) as it cannot be copy-pasted when you simply paste it as text

**Reply by Vfriday** [reply]
I have try to edit it but not getting the expected result.

**Reply by johanlofberg** [reply]
Dehomogenizing by X>=I shows that the problem is infeasible

**Reply by johanlofberg** [reply]
Your problem is homogenuous hence zero is a trivial solution. You are replacing strict conditions with non-strict without understand the implications, and must thus dehomogenize in other way. Final section here https://yalmip.github.io/badsdps and discussed in SDP tutorial too

**Reply by Vfriday** [reply]
When I dehomogenize X>=eps*I, i got zero solution

**Reply by johanlofberg** [reply]
Well, eps, if you mean matlab default eps, is basically zero to all numerical solvers, so the solver is simply saying that that problem is feasible (with the 0 solution) up to numerical tolerances. However, since you know the problem is homogeneous, it should be feasible for any dehomogenization, and since it isn't, it is structurally infeasible

It's like trying to solve the problem find x such that x>0 and x<0. Trivially infeasible. Implementing it with non-strict inequalities leads to x>=0, x<=0 which of course leads to infeasibility. As it is homogenuous, any dehomogenization should work so x>=1,x<=-1 is thus used and fails. Using x>eps, x<-eps could easily lead to the solver being ok with returning the solution 0 (although in this trivial case most solver would detect this as infeasible in the presolve)

**Reply by Vfriday** [reply]
I will like to ask one more question sir, the continuous-time model worked well with the yalmip optimization but experiencing this problem with the discrete-time model though I have applied several discretization methods , what other reasons do you think could be responsible for this challenge?

**Reply by johanlofberg** [reply]
impossible to answer (assuming you've found/derived the correct theory for the discrete version)

**Reply by Vfriday** [reply]
My mistake was trying to place the poles beyond the unit circle.
